### ISSUE TITLE
Avoid putting a bos token before answer ？

### ISSUE THREAD

**Issue by Alex-Songs** [ddlBoJack/SLAM-LLM#90]
Hello,
   I don't quite understand why bos is not added here "(example = prompt + answer  # FIX(MZY): avoid putting a bos token before answer.)". How can autoregressive training be implemented without adding bos?

**Comment by Alex-Songs** [comment]
adding <bos>

**Comment by ddlBoJack** [comment]
Hi, if you call `self.tokenizer.encode()` function, a bos token will be added before the sequence you encode. We only want to add the bos token before "USER: {prompt}\n ASSISTANT:{answer}", rather than before "{answer}".

**Comment by ddlBoJack** [comment]
Hi, if you call `self.tokenizer.encode()` function, a bos token will be added before the sequence you encode. We only want to add the bos token before "USER: {prompt}\n ASSISTANT:{answer}", rather than before "{answer}".

**Comment by Alex-Songs** [comment]
Thank you, I have another question. Why is the attention mask all True except for the padding part, and where is the casual mask required by the autoregressive model constructed? @ddlBoJack

**Comment by ddlBoJack** [comment]
All the tokens in the sequence are masked casually. Pay attention that -100 is the special token for ignoring computing loss, which is set in the label sequence.

**Comment by Alex-Songs** [comment]
@ddlBoJack However, I noticed in speech_dataset.py that example_mask is all true: example_mask = example_ids.ge(-1)  # FIX(GZF): [True,True,True,True]. This part is assigned to attention_mask.

**Comment by ddlBoJack** [comment]
Llama model will tackle the mask. A triangular matrix will be generated inside the forward pipeline. So here attention_mask is the padding mask.

**Comment by Alex-Songs** [comment]
Thanks
